Automattic/jetpack
# AI Assistant: enable backend prompts when the beta feature is available

Part of #28569.

~~Depends on #33017. Will need a rebase after it to have the beta flag code available.~~ Rebased.

~~Depends on D121522-code. Test in pair with it.~~ Merged and deployed.

## Proposed changes:

* Use a beta feature to decide between using client or backend prompts
* When the beta feature is available, short-circuit the original prompt generation to use backend prompt messages
* After a lot of back-and-forth, I had to bring back the idea of a message that contains the relevant content of the request, and I copied the exact same rule for the relevant content that the current implementation uses (check the `switch` block on the `backend-prompt.ts` file)
* This PR is enabling all the prompt types at once but hidding it as beta

As follow-ups on this, after finishing the testing, we may need to:

* drop the beta feature check
* remove the old code so the files become cleaner to understand

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* ~~Test in pair with D121522-code~~
* ~~Make sure your test site is pointing to the `public-api`in your sandbox, and your sandbox is running the code above~~
* Make sure you have beta extensions enabled on your test site
* Go to the block editor and open the browser console
* Enable debug messages by running the following command; refresh the window after it:

```
localStorage.setItem( 'debug', '*' );
```

* Insert an AI Assistant block and ask for some completion (example: a haiku about code)
* Notice the `jetpack-ai-client:ask-question` log message
   * the message will follow the new format
   * if the post has some content, confirm the correct relevant content is present
   * confirm that the `context.for` property contains the right target action

<img width="500" alt="Screenshot 2023-09-12 at 18 48 14" src="https://github.com/Automattic/jetpack/assets/6760046/ab3c2c51-0249-4151-9ed6-53e70dc71c84">

* Ask follow-up actions on the response and notice the history of messages being sent:

<img width="500" alt="Screenshot 2023-09-12 at 18 48 43" src="https://github.com/Automattic/jetpack/assets/6760046/a9cd1f88-8dc1-4040-8017-e102505ad978">

---

* Confirm that the actions on the assistant work as expected (example of "Generate title"):

<img width="500" alt="Screenshot 2023-09-12 at 18 51 36" src="https://github.com/Automattic/jetpack/assets/6760046/373c07c9-0c97-4b82-b9c1-3a6bd6335e61">

<img width="500" alt="Screenshot 2023-09-12 at 18 51 41" src="https://github.com/Automattic/jetpack/assets/6760046/d9cb6df7-39ad-4ab8-8d81-0b40ef124e8a">

---

* Confirm that the actions on the paragraph extension work as expected by inserting some paragraph with content and using the extension options (or typing a custom prompt):

<img width="500" alt="Screenshot 2023-09-12 at 18 51 02" src="https://github.com/Automattic/jetpack/assets/6760046/ccb87098-c455-4754-a3ad-b2b034496d75">

<img width="500" alt="Screenshot 2023-09-12 at 18 51 08" src="https://github.com/Automattic/jetpack/assets/6760046/a92d3676-e3d9-4b63-81e0-a54376f5a0cf">

---

* Check the logs on logstash to confirm that the prompts are being generated correctly:

<img width="500" alt="Screenshot 2023-09-12 at 18 56 14" src="https://github.com/Automattic/jetpack/assets/6760046/9befe678-3883-4087-a8c3-abe950ae0307">

<img width="500" alt="Screenshot 2023-09-12 at 18 56 45" src="https://github.com/Automattic/jetpack/assets/6760046/7b5095a4-bb77-422e-9e04-c98bd689b89f">

* Try to cover as much features as possible, looking for regressions or unexpected behavior

---

* Disable the beta extensions and try to use the assistant
* It should behave as usual, sending the prompts from the client:

<img width="600" alt="Screenshot 2023-09-13 at 15 51 26" src="https://github.com/Automattic/jetpack/assets/6760046/ee872d20-0a5b-49f4-bcef-a8b913ef22ec">

<img width="600" alt="Screenshot 2023-09-13 at 15 51 45" src="https://github.com/Automattic/jetpack/assets/6760046/78565dc0-f178-48ea-bd91-b256804381b1">

<img width="600" alt="Screenshot 2023-09-13 at 15 52 01" src="https://github.com/Automattic/jetpack/assets/6760046/6768da35-7223-47a9-8329-15f8e720c9b2">

* The client prompts should work as expected without any regression

---

* Make sure you tested on Simple, Jetpack, Atomic and P2
